### PR TITLE
Add `exports` field to `Config`

### PR DESCRIPTION
**Pull request by adambratschikaye**
Add an optional `exports` field to `Config` which describes exactly which exports the generated module should have.

This is similar to the solution described [here](https://github.com/bytecodealliance/wasm-tools/issues/1405#issuecomment-1934328598).

### CODE REVIEW

**Comment by adambratschikaye** [comment]
This is failing on a seemingly unrelated test for `wit-parser`, but I also get that same failure on `main` commit https://github.com/bytecodealliance/wasm-tools/commit/2573bb28cc897d64ce33739e3b46ac63c9588267.

**Comment by alexcrichton** [comment]
Ah yeah no worries on the wit-parser failure, that's due to https://github.com/bytecodealliance/wasm-tools/pull/1419

File: crates/wasm-smith/tests/exports.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                }
+            } else if let wasmparser::Payload::GlobalSection(rdr) = payload {
+                for global in rdr {
+                    let global = global.expect("Failed to read global");
+                    globals.push(global.ty);
+                }
+            } else if let wasmparser::Payload::ExportSection(rdr) = payload {
+                exports = rdr.into_iter().collect::<Result<_, _>>().unwrap();
+            }
+        }
```
Comment (alexcrichton) [inline crates/wasm-smith/tests/exports.rs:206]: One thing I might recommend is to use the [result of validation](https://docs.rs/wasmparser/latest/wasmparser/types/struct.Types.html) which should avoid the need to manually walk over these sections and record the results.

Comment (alexcrichton) [reply crates/wasm-smith/tests/exports.rs:206]: You might also be able to skip the `expected_exports` field and instead run the validator over the `export_module` field and then assert everything there is also present in the wasm-smith output

**Review by alexcrichton — COMMENTED** [review]
Thanks again for this! Overall everything looks reasonable to me, but if you're up for it could you look into using `wasmparser::types::Types`, the result of validation, instead of parsing the input module directly? I left some comments below on the tests but I think this is also applicable to where the example module is parsed and read. I think that might help simplify some of the logic here where you would only need to buffer up the results of the `ExportSection` and then that could be passed to `Types` directly to get all the type information (via [`entity_type_from_export`](https://docs.rs/wasmparser/latest/wasmparser/types/struct.TypesRef.html#method.entity_type_from_export) and [`EntityType`](https://docs.rs/wasmparser/latest/wasmparser/types/enum.EntityType.html)).

If you feel though that it doesn't end up simplifying much and it's about as complex I'm also happy to land this as-is.

**Comment by alexcrichton** [comment]
Also, as a minor note, after https://github.com/bytecodealliance/wasm-tools/pull/1416 we're changing a bit how CI works on this repository, so CI won't fully finish on this PR just yet until it's rebased. After rebasing on `main` though it should be good to go.

File: crates/wasm-smith/tests/exports.rs (hunk 1 above)
Comment (adambratschikaye) [reply crates/wasm-smith/tests/exports.rs]: Thanks, I switched to using `Validator` to get `Types` in both the test and the implementation.

**Review by alexcrichton — APPROVED** [review]
Thank you again for this, it's very much appreciated!

**Pull request merged**